### PR TITLE
Matrix improvements

### PR DESCRIPTION
**Pull request by isuruf**
@certik, @moorepants

### CODE REVIEW

**Comment by isuruf** [comment]
Ready for review

File: symengine/dense_matrix.cpp [hunk 1]
```diff
@@ -175,14 +181,75 @@ void jacobian(const DenseMatrix &A, const DenseMatrix &x, DenseMatrix &result)
...
+            "Use sjacobian for SymPy style differentiation");
+    }
+}
+
+void sjacobian(const DenseMatrix &A, const DenseMatrix &x, DenseMatrix &result)
+{
+    SYMENGINE_ASSERT(A.col_ == 1);
+    SYMENGINE_ASSERT(x.col_ == 1);
+    SYMENGINE_ASSERT(A.row_ == result.nrows() and x.row_ == result.ncols());
+#pragma omp parallel for
```
Comment (moorepants) [inline symengine/dense_matrix.cpp]: I like this. I assume this parallelizes the jacobian computation.

**Comment by moorepants** [comment]
I'm an amature at C++ so reading the uncommented code isn't that easy for me. Can you explain in words what you have added here? What is the purpose of the "s" versions of the functions?

File: symengine/derivative.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+//! Differentiation w.r.t symbols
+RCP<const Basic> diff(const RCP<const Basic> &arg, const RCP<const Symbol> &x);
+
+//! SymPy style differentiation w.r.t non-symbols and symbols
+RCP<const Basic> sdiff(const RCP<const Basic> &arg, const RCP<const Basic> &x);
+
+} // namespace SymEngine
+
+#endif // SYMENGINE_DERIVATIVE_H
```
Comment (moorepants) [inline symengine/derivative.h]: eol

File: symengine/dense_matrix.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/dense_matrix.cpp]: Yes

File: symengine/derivative.cpp [hunk 3]
```diff
@@ -590,4 +591,20 @@ class DiffImplementation
 #include "symengine/type_codes.inc"
 #undef SYMENGINE_ENUM
 
+RCP<const Basic> diff(const RCP<const Basic> &arg, const RCP<const Symbol> &x) {
+    return arg->diff(x);
+}
+
+//! SymPy style differentiation for non-symbol variables
+// Since SymPy's differentiation makes no sense mathematically, it is
+// defined separately here for compatibility
```
Comment (isuruf) [inline symengine/derivative.cpp]: Difference is mentioned here. Basically `diff` takes only a `Symbol` as the w.r.t argument. `sdiff` takes non-symbols as well and differentiate it as in SymPy. (Substitute with a dummy symbol, differentiate and then substitute back)

File: symengine/dense_matrix.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/dense_matrix.cpp]: Only when SymEngine is compiled with `WITH_OPENMP` and the compiler supports openmp standard

**Comment by isuruf** [comment]
Fixed merge conflicts

**Comment by isuruf** [comment]
@certik, ping

File: symengine/dense_matrix.cpp [hunk 4]
```diff
@@ -12,6 +12,11 @@ DenseMatrix::DenseMatrix()
 {
 }
 
+DenseMatrix::DenseMatrix(const DenseMatrix &x)
+    : m_(x.m_), row_(x.row_), col_(x.col_)
+{
+}
+
```
Comment (certik) [inline symengine/dense_matrix.cpp]: Put this into the header file for efficiency reasons.

**Comment by certik** [comment]
I think this looks good. Merging. We should put all the constructors into header files.

**Pull request merged**